### PR TITLE
KVDB Manager implementation

### PR DESCRIPTION
**Pull request by LucioDonda**
|Related issue|
|---|
|#12819 #12822 |

## Description
This PR implements a KVDB Manager capable to create, delete, and handle multiple Opened Key-Value DBs.
Wrapping KVDB methods inside a specific class allows it to be used with several DB files 

It also adds support for legacy CDB files that are now converted into KVDB files.

Also, UT for different usages of the library where implemented.

### CODE REVIEW

File: src/engine/source/kvdb/src/kvdb.hpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+class KVDB
+{
+public:
+    KVDB(const std::string &dbName, const std::string &folder);
+    KVDB() = default;
+    KVDB(KVDB const &) = delete;
+    void operator=(KVDB const &) = delete;
+    ~KVDB();
+
+    std::string &getName()
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdb.hpp:22]: could mark the 'get' funcs as const

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        return false;
+    }
+
+    std::string line;
+    while (getline(CDBfile, line))
+    {
+        line.erase(remove_if(line.begin(), line.end(), isspace), line.end());
+        auto KV = utils::string::split(line, ':');
+        kvdb.write(KV[0], KV[1]);
+    }
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdbManager.cpp:134]: with c++17 we can do
```suggestion
    while (std::string line; getline(CDBfile, line))
    {
        line.erase(remove_if(line.begin(), line.end(), isspace), line.end());
        auto KV = utils::string::split(line, ':');
        kvdb.write(KV[0], KV[1]);
    }
```
This way the string has the explicit 'lifetime' it needs

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    {
+        // TODO Log Error
+        CDBfile.close();
+        return false;
+    }
+    KVDB &kvdb = getDB(path.stem());
+    if (kvdb.getState() != KVDB::State::Open)
+    {
+        // TODO Log Error
+        CDBfile.close();
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdbManager.cpp:124]: same as above

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        return false;
+    }
+
+    std::string line;
+    while (getline(CDBfile, line))
+    {
+        line.erase(remove_if(line.begin(), line.end(), isspace), line.end());
+        auto KV = utils::string::split(line, ':');
+        kvdb.write(KV[0], KV[1]);
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdbManager.cpp:133]: should check that the return is not empty or this will crash (for example with a format error)

**Review by san1526 — CHANGES_REQUESTED** [review]

File: src/engine/source/kvdb/include/kvdb/kvdbManager.hpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+public:
+    static KVDBManager &getInstance()
+    {
+        static KVDBManager instance;
+        return instance;
+    }
+    KVDBManager(KVDBManager const &) = delete;
+    void operator=(KVDBManager const &) = delete;
+    bool createDB(const std::string &Name, bool replace = true);
```
Comment (san1526) [inline src/engine/source/kvdb/include/kvdb/kvdbManager.hpp:28]: `createDB` should return the created db or we are going to have the pattern below everywhere
```
KVDBManager::getInstance().createDb("name");
KVDBManager::getInstance().getDb("name");
```

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 6]
```diff
@@ -1,561 +1,511 @@
...
  *
+ * @param dbName name of the DB
+ * @param folder where the DB will be stored
  */
-void UpdateColumnFamiliesList() {
-
-    std::vector<std::string> pColumn_families;
-
-    DB::ListColumnFamilies(DBOptions(), kKvDbPath, &pColumn_families);
+KVDB::KVDB(const std::string &dbName, const std::string &folder)
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdb.cpp:21]: use the member initialization list

File: src/engine/source/kvdb/include/kvdb/kvdbManager.hpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+class KVDBManager
+{
+    KVDBManager();
+    ~KVDBManager();
+    const std::string FOLDER = "/var/ossec/queue/db/kvdb/";
+    using DBMap = std::unordered_map<std::string, KVDB *>;
+    DBMap available_kvdbs;
+    bool addDB(KVDB *DB);
+    KVDB invalidDB = KVDB(); // TODO: Make this static?
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/include/kvdb/kvdbManager.hpp]: done

File: src/engine/source/kvdb/include/kvdb/kvdbManager.hpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+#include "kvdb.hpp"
+
+class KVDBManager
+{
+    KVDBManager();
+    ~KVDBManager();
+    const std::string FOLDER = "/var/ossec/queue/db/kvdb/";
+    using DBMap = std::unordered_map<std::string, KVDB *>;
+    DBMap available_kvdbs;
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/include/kvdb/kvdbManager.hpp]: done, changed to `m_availableKVDBs`

File: src/engine/source/kvdb/include/kvdb/kvdbManager.hpp [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    bool addDB(KVDB *DB);
+    KVDB invalidDB = KVDB(); // TODO: Make this static?
+
+public:
+    static KVDBManager &getInstance()
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/include/kvdb/kvdbManager.hpp]: `get` sounds cleaner.

File: src/engine/source/kvdb/include/kvdb/kvdbManager.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/include/kvdb/kvdbManager.hpp]: done

File: src/engine/source/kvdb/include/kvdb/kvdbManager.hpp [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        static KVDBManager instance;
+        return instance;
+    }
+    KVDBManager(KVDBManager const &) = delete;
+    void operator=(KVDBManager const &) = delete;
+    bool createDB(const std::string &Name, bool replace = true);
+    bool createDBfromCDB(const std::filesystem::path &path,
+                         bool replace = true);
+    bool DeleteDB(const std::string &name);
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/include/kvdb/kvdbManager.hpp]: done

File: src/engine/source/kvdb/include/kvdb/kvdbManager.hpp [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    KVDB &getDB(const std::string &Name);
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/include/kvdb/kvdbManager.hpp]: done

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+                     // or if receives a CFHandler and a CFDescriptor vector.
+    if (!s.ok())
+    {
+        // LOG(WARNING) << "[" << __func__ << "]" << " couldn't close db file,
+        // error: " << s.ToString() << std::endl;
+        return false;
+    }
+
+    KVDB *kvdb = new KVDB(name, FOLDER);
+    if (!addDB(kvdb))
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: yup, good catch.

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+        WAZUH_LOG_ERROR("Database [{}] isn´t handled by KVDB manager", name);
+        ret = false;
+    }
+
+    return ret;
+}
+
+bool KVDBManager::addDB(KVDB *DB)
+{
+    if (available_kvdbs.find(DB->getName()) == available_kvdbs.end())
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: check with the new implementation.

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+    createOptions.IncreaseParallelism();
+    createOptions.OptimizeLevelStyleCompaction();
+    createOptions.create_if_missing = true;
+    createOptions.error_if_exists = !replace;
+
+    ROCKSDB::DB *db;
+    ROCKSDB::Status s = ROCKSDB::DB::Open(createOptions, FOLDER + name, &db);
+    if (!s.ok())
+    {
+        // LOG(ERROR) << "[" << __func__ << "]" << " couldn't open db file,
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: done

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+                    // std::filesystem::remove(cdbfile.path())
+                }
+            }
+        }
+    }
+}
+
+KVDBManager::~KVDBManager()
+{
+    for (DBMap::iterator it = available_kvdbs.begin();
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: done

File: src/engine/source/kvdb/CMakeLists.txt [hunk 16]
```diff
@@ -33,29 +33,23 @@ CPMAddPackage(
...
-)
-
 # Add this dependencies to parent CMakeLists scope
 set(EXTERNAL_DEPS ${EXTERNAL_DEPS} PARENT_SCOPE)
 
 ####################################################################################################
 # Sources - Includes
 ####################################################################################################
-add_library(kvdb ${ENGINE_KVDB_SOURCE_DIR}/kvdb.cpp)
+add_library(kvdb
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/CMakeLists.txt]: done, modified to STATIC

File: src/engine/source/kvdb/CMakeLists.txt [hunk 17]
```diff
@@ -33,29 +33,23 @@ CPMAddPackage(
...
+    ${ENGINE_KVDB_SOURCE_DIR}/kvdb.hpp
+    )
+
 
-target_link_libraries(kvdb PRIVATE rocksdb "glog::glog")
+target_link_libraries(kvdb PRIVATE rocksdb base)
 
 target_include_directories(kvdb PUBLIC
     ${ENGINE_KVDB_INCLUDE_DIR}
+    ${ENGINE_KVDB_SOURCE_DIR}
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/CMakeLists.txt:39]: Being ENGINE_KVDB_SOURCE_DIR private I'm having errors trying to include kvdb.hpp in kvdbManager

File: src/engine/source/kvdb/src/kvdb.hpp (hunk 1 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.hpp]: done!

File: src/engine/source/kvdb/src/kvdb.hpp [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+    State getState()
+    {
+        return m_state;
+    }
+
+    bool createColumn(const std::string &columnName);
+
+    bool deleteColumn(const std::string &columnName =
+                          ROCKSDB_NAMESPACE::kDefaultColumnFamilyName);
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.hpp]: done

File: src/engine/source/kvdb/src/kvdb.hpp [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+#define _KVDB_H
+
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+#include "rocksdb/db.h"
+#include "rocksdb/utilities/transaction_db.h"
+
+namespace ROCKSDB = ROCKSDB_NAMESPACE;
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.hpp]: ok

File: src/engine/source/kvdb/src/kvdb.hpp [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+    };
+
+    State getState()
+    {
+        return m_state;
+    }
+
+    bool createColumn(const std::string &columnName);
+
+    bool deleteColumn(const std::string &columnName =
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.hpp]: the idea behind this concept was to show the user the possibility of not using Column Families (in fact rocksdb is using the default one) so, if this is changed then the user will be forced to delete or clean a "default" CF, which he did not create.
This was Moved to a TODO, meanwhile to avid using rocksdb variable in the client this was changed to a own define (DEFAULT_CF_NAME)

File: src/engine/source/kvdb/src/kvdb.hpp [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+                          ROCKSDB_NAMESPACE::kDefaultColumnFamilyName);
+
+    bool cleanColumn(const std::string &columnName =
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.hpp]: Moved to a TODO

File: src/engine/source/kvdb/src/kvdb.hpp [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+               const std::string &value,
+               const std::string &columnName =
+                   ROCKSDB_NAMESPACE::kDefaultColumnFamilyName);
+
+    bool writeToTransaction(
+        const std::vector<std::pair<std::string, std::string>> pairsVector,
+        const std::string &columnName =
+            ROCKSDB_NAMESPACE::kDefaultColumnFamilyName);
+
+    bool exist(const std::string &key,
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.hpp]: Moved to a TODO

File: src/engine/source/kvdb/src/kvdb.hpp [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+    std::string read(const std::string &key,
+                     const std::string &columnName =
+                         ROCKSDB_NAMESPACE::kDefaultColumnFamilyName);
+
+    bool readPinned(const std::string &key,
+                    std::string &val,
+                    const std::string &columnName =
+                        ROCKSDB_NAMESPACE::kDefaultColumnFamilyName);
+
+    bool deleteKey(const std::string &key,
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.hpp:68]: Moved to a TODO

File: src/engine/source/kvdb/src/kvdb.hpp [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+
+private:
+    std::string m_name = "Invalid";
+    std::string m_path;
+    State m_state = State::Invalid;
+
+    ROCKSDB::DB *m_db;
+    ROCKSDB::TransactionDB *m_txndb;
+
+    struct Option
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.hpp]: done -> move to a static const in the .cpp

File: src/engine/test/source/kvdb/CMakeLists.txt [hunk 25]
```diff
@@ -1,4 +1,3 @@
-
 CPMAddPackage(
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/test/source/kvdb/CMakeLists.txt]: done

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 26]
```diff
@@ -0,0 +1,181 @@
...
+        available_kvdbs.erase(it);
+    }
+    else
+    {
+        WAZUH_LOG_ERROR("Database [{}] isn´t handled by KVDB manager", name);
+        ret = false;
+    }
+
+    return ret;
+}
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: great!

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+    return true;
+}
+
+bool KVDBManager::DeleteDB(const std::string &name)
+{
+    bool ret = true;
+    auto it = available_kvdbs.find(name);
+    if (it != available_kvdbs.end())
+    {
+        ret = it->second->destroy();
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: I'm changing de method in kvdb for, at least, setting the state of the DB to error and communicating the user that it has an error.

File: src/engine/source/kvdb/src/kvdbManager.cpp (hunk 2 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: can't make it work with the **while()** clause (but **if()** has no problem) 
```
 error: expected initializer before ‘;’ token
  116 |     while (std::string line; getline(CDBfile, line))
      |                            ^
```

File: src/engine/source/kvdb/src/kvdbManager.cpp (hunk 3 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: done

File: src/engine/source/kvdb/src/kvdbManager.cpp (hunk 4 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: right, changed!

File: src/engine/source/kvdb/CMakeLists.txt [hunk 28]
```diff
@@ -33,29 +33,23 @@ CPMAddPackage(
...
-      "WITH_GTEST OFF"
-      "BUILD_SHARED_LIBS OFF"
-      "WITH_GFLAGS OFF"
-      "WITH_UNWIND OFF"
-)
-
 # Add this dependencies to parent CMakeLists scope
 set(EXTERNAL_DEPS ${EXTERNAL_DEPS} PARENT_SCOPE)
 
 ####################################################################################################
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/CMakeLists.txt:25]: done

File: src/engine/source/kvdb/include/kvdb/kvdbManager.hpp (hunk 5 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/include/kvdb/kvdbManager.hpp]: I have change the return value in order to be able to use
`KVDB &newKvdb = kvdbManager.createDB("NEW_DB");`
But theres also the option of using it as
```
    kvdbManager.deleteDB("NEW_DB");
    KVDB &deletedKvdb = kvdbManager.getDB("NEW_DB");
```

File: src/engine/source/kvdb/src/kvdb.cpp (hunk 6 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.cpp:30]: done!

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+    if (!CDBfile.is_open())
+    {
+        // TODO Log Error
+        return false;
+    }
+
+    if (!createDB(path.stem(), replace))
+    {
+        // TODO Log Error
+        CDBfile.close();
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: done

File: src/engine/source/kvdb/CMakeLists.txt (hunk 17 above)
[the remark this answers is not in the document]
Comment (san1526) [reply src/engine/source/kvdb/CMakeLists.txt:39]: being that kvdb.hpp is part of our interface it should actually be public and be in the include folder instead

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 30]
```diff
@@ -1,561 +1,529 @@
...
+        if (!s.ok())
+        {
+            WAZUH_LOG_ERROR("couldn't insert value into CF, error: [{}]",
+                            s.ToString());
+            return false;
+        }
 
-    if(CFIndexInAvailableArray(column_family_name)) {
-        LOG(INFO) << "[" << __func__ << "]" << " cant create a family column already present" << std::endl;
+        WAZUH_LOG_INFO("value insertion OK [{},{}] into CF name : [{}]",
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdb.cpp:177]: this should probably be a DEBUG log instead

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 31]
```diff
@@ -1,561 +1,529 @@
...
-            }
-
+    if (m_state == State::Open)
+    {
+        // Get CF handle
+        rocksdb::Status s =
+            m_db->Get(kOptions.read, cfh->second, rocksdb::Slice(key), &value);
+        if (s.ok())
+        {
+            WAZUH_LOG_INFO("value obtained OK [{},{}] into CF name : [{}]",
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdb.cpp:216]: should probably be DEBUG log

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 32]
```diff
@@ -1,561 +1,529 @@
...
+            m_db->Delete(kOptions.write, cfh->second, rocksdb::Slice(key));
+        if (s.ok())
+        {
+            WAZUH_LOG_INFO("key deleted OK [{}]", key);
+            return true;
         }
-        else {
-            LOG(ERROR) << "[" << __func__ << "]" << " couldn't open db file, error: " << s.ToString() << std::endl;
-            result = false;
+        else
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdb.cpp:261]: avoid the else's in arrow pattern by just flipping the condition and doing an early return

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 33]
```diff
@@ -1,561 +1,529 @@
...
-                        }
-                    }
+    if (m_state == State::Open)
+    {
+        rocksdb::Transaction *txn = m_txndb->BeginTransaction(kOptions.write);
+        if (txn)
+        {
+            for (auto pair : pairsVector)
+            {
+                std::string const key = pair.first;
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdb.cpp:418]: should avoid copying here. You can use structured binding in the for loop

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 34]
```diff
@@ -1,561 +1,529 @@
...
         }
-        delete txn_db;
     }
-    else {
-        LOG(ERROR) << "[" << __func__ << "]" << " can't delete a FC that doesn't exist" << std::endl;
-        result = false;
+    else
+    {
+        WAZUH_LOG_ERROR("DB should be open for execution");
+        return false;
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdb.cpp:463]: i find this function a bit hard to follow. Consider refactoring like this (this is the way to avoid the arrow pattern mentioned above)
```
bool KVDB::writeToTransaction(
    const std::vector<std::pair<std::string, std::string>> pairsVector,
    const std::string &columnName)
{
    if (m_state != State::Open)
    {
        WAZUH_LOG_ERROR("DB should be open for execution");
        return false;
    }

    if (!pairsVector.size())
    {
        WAZUH_LOG_ERROR(
            "couldn't write transaction to DB, nedd at least 1 element");
        return false;
    }

    CFHMap::const_iterator cfh = m_CFHandlesMap.find(columnName);
    if (cfh == m_CFHandlesMap.end())
    {
        WAZUH_LOG_ERROR("couldn't write transaction to DB unknown column name");
        return false;
    }

    auto txn = std::unique_ptr<rocksdb::Transaction>(m_txndb->BeginTransaction(kOptions.write));
    if (!txn)
    {
        WAZUH_LOG_ERROR("couldn't begin in transaction");
        return false;
    }

    for (auto [key, value] : pairsVector)
    {
        if (key.empty())
        {
            WAZUH_LOG_ERROR("can't write to a Transaction to a family "
                            "column with no key.");
            return false;
        }
        // Write a key in this transaction
        rocksdb::Status s = txn->Put(cfh->second, key, value);
        if (!s.ok())
        {
            WAZUH_LOG_ERROR("couldn't execute Put in transaction "
                            "-breaking loop-, error: [{}]",
                            s.ToString());
            return false;
        }
    }

    rocksdb::Status s = txn->Commit();
    if (!s.ok())
    {
        WAZUH_LOG_ERROR("couldn't commit in transaction, error: [{}]",
                        s.ToString());
        return false;
    }

    WAZUH_LOG_DEBUG("transaction commited OK");
    return true;
}
```

**Review by san1526 — CHANGES_REQUESTED** [review]

File: src/engine/source/kvdb/src/kvdb.cpp (hunk 34 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.cpp]: awesome!

File: src/engine/source/kvdb/src/kvdb.cpp (hunk 32 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.cpp]: right, done!

File: src/engine/source/kvdb/src/kvdb.cpp (hunk 31 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.cpp]: ok

File: src/engine/source/kvdb/src/kvdb.cpp (hunk 30 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.cpp]: done

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 35]
```diff
@@ -1,561 +1,529 @@
...
         return false;
     }
 
-    dbOptions.IncreaseParallelism();
-    dbOptions.create_if_missing = true;
-    dbOptions.error_if_exists = false;
-    dbOptions.create_missing_column_families = true;
-
-    UpdateColumnFamiliesList();
+    if (m_state == State::Open)
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.cpp]: done!

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 36]
```diff
@@ -1,561 +1,529 @@
...
+                WAZUH_LOG_INFO("transaction commited OK");
+                delete txn;
+                return true;
             }
-            else {
-                LOG(ERROR) << "[" << __func__ << "]" << " couldn't begin transaction, error: " << s.code() << std::endl;
-                result =  false;
+            else
+            {
+                WAZUH_LOG_ERROR("couldn't commit in transaction, error: [{}]",
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.cpp]: indeed.

File: src/engine/source/kvdb/src/kvdb.hpp [hunk 37]
```diff
@@ -0,0 +1,86 @@
...
+    enum class State
+    {
+        Closed,
+        Open,
+        Locked,
+        Error,
+        Invalid,
+    };
+
+    State getState()
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.hpp]: ok

File: src/engine/source/kvdb/src/kvdb.hpp [hunk 38]
```diff
@@ -0,0 +1,86 @@
...
+#define _KVDB_H
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "rocksdb/db.h"
+#include "rocksdb/utilities/transaction_db.h"
+
+#define DEFAULT_CF_NAME "default"
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.hpp]: done like this `constexpr static char* DEFAULT_CF_NAME {"default"};`
can this be modify in order to avoid warnings on compile time?
```
warning: ISO C++ forbids converting a string constant to ‘char*’ [-Wwrite-strings]
   71 |     constexpr static char* DEFAULT_CF_NAME {"default"};

```

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 39]
```diff
@@ -0,0 +1,170 @@
...
+bool KVDBManager::addDB(std::string name, std::string folder)
+{
+    KVDB *kvdb;
+    if (m_availableKVDBs.find(name) == m_availableKVDBs.end())
+    {
+        kvdb = new KVDB(name, folder);
+        m_availableKVDBs[name] = std::unique_ptr<KVDB>(kvdb);
+        return true;
+    }
+    else
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: right.

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 40]
```diff
@@ -0,0 +1,170 @@
...
+        WAZUH_LOG_ERROR("Database [{}] isn´t handled by KVDB manager", name);
+        return false;
+    }
+    bool ret = it->second->destroy();
+    it->second.release();
+    m_availableKVDBs.erase(it);
+    return ret;
+}
+
+bool KVDBManager::addDB(std::string name, std::string folder)
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: done

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 41]
```diff
@@ -1,561 +1,509 @@
...
-                LOG(ERROR) << "[" << __func__ << "]" << " couldn't begin transaction, error: " << s.code() << std::endl;
-                result =  false;
-            }
+    for (auto & [key, value] : pairsVector)
+    {
+        if (key.empty())
+        {
+            WAZUH_LOG_ERROR("can't write to a Transaction to a family "
+                            "column with no key.");
+            return false;
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdb.cpp:421]: returning here is a memory leak

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 42]
```diff
@@ -1,561 +1,509 @@
...
+        if (s.ok())
+        {
+            continue;
         }
-        else {
-            LOG(ERROR) << "[" << __func__ << "]" << " couldn't Open DB, error: " << s.code() << std::endl;
-            result = false;
+        else
+        {
+            WAZUH_LOG_ERROR("couldn't execute Put in transaction "
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdb.cpp:431]: returning here is a memory leak

This is why the suggestion to just change it to the 'smart' pointer instead

File: src/engine/source/kvdb/src/kvdb.cpp [hunk 43]
```diff
@@ -1,561 +1,509 @@
...
     }
 
-    return result;
+    rocksdb::PinnableSlice pinnable_val;
+    rocksdb::Status s = m_db->Get(
+        kOptions.read, cfh->second, rocksdb::Slice(key), &pinnable_val);
+    if (s.ok())
+    {
+        value = pinnable_val.ToString();
+        WAZUH_LOG_INFO("read pinned value OK [{},{}]", key, value);
```
Comment (san1526) [inline src/engine/source/kvdb/src/kvdb.cpp:501]: probably should be a debug log

**Review by san1526 — CHANGES_REQUESTED** [review]

File: src/engine/source/kvdb/src/kvdb.cpp (hunk 42 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.cpp:427]: this should be fixed [here](https://github.com/wazuh/wazuh/issues/12935)

File: src/engine/source/kvdb/src/kvdb.cpp (hunk 41 above)
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdb.cpp:421]: this should be fixed [here](https://github.com/wazuh/wazuh/issues/12935)

File: src/engine/source/kvdb/src/kvdbManager.cpp [hunk 44]
```diff
@@ -0,0 +1,170 @@
...
+    }
+    bool ret = it->second->destroy();
+    it->second.release();
+    m_availableKVDBs.erase(it);
+    return ret;
+}
+
+bool KVDBManager::addDB(const std::string &name,const std::string &folder)
+{
+    KVDB *kvdb;
```
[the remark this answers is not in the document]
Comment (LucioDonda) [reply src/engine/source/kvdb/src/kvdbManager.cpp]: deleted

**Review by san1526 — APPROVED** [review]
If we have a fix for the mem leaks this is good to go
